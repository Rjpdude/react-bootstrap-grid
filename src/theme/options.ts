export interface ThemeOptions {
  /**
   * The `margin` and `padding` in pixels for all grid elements.
   *
   * @default 15
   */
  spacing: number
  /**
   * The maximum width in pixels of the small breakpoint container.
   *
   * @default 576
   */
  sm: number
  /**
   * The width in pixels to assign to the small breakpoint container.
   *
   * @default 540
   */
  smWidth: number
  /**
   * The maximum width in pixels of the medium breakpoint container.
   *
   * @default 768
   */
  md: number
  /**
   * The width in pixels to assign to the medium breakpoint container.
   *
   * @default 720
   */
  mdWidth: number
  /**
   * The maximum width in pixels of the large breakpoint container.
   *
   * @default 992
   */
  lg: number
  /**
   * The width in pixels to assign to the large breakpoint container.
   *
   * @default 960
   */
  lgWidth: number
  /**
   * The maximum width in pixels of the extra-large breakpoint container.
   *
   * @default 1200
   */
  xl: number
  /**
   * The width in pixels to assign to the extra-large breakpoint container.
   *
   * @default 1140
   */
  xlWidth: number
}

export function getDefaultTheme(): ThemeOptions {
  return {
    spacing: 15,
    sm: 576,
    smWidth: 540,
    md: 768,
    mdWidth: 720,
    lg: 992,
    lgWidth: 960,
    xl: 1200,
    xlWidth: 1140
  }
}
