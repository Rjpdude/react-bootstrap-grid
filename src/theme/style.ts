import { css, createGlobalStyle } from 'styled-components'

export const Global = css`
  html {
    box-sizing: border-box;
    -ms-overflow-style: scrollbar;
  }

  *,
  *::before,
  *::after {
    box-sizing: inherit;
  }
`

export const GridThemeStyle = createGlobalStyle`
  ${Global}
`
