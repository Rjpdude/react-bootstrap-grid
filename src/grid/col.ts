import { CSSObject } from 'styled-components'
import { ThemeOptions } from '../theme/options'
import { GridElement } from '../element/interface'
import { configureElement } from '../element/element'
import { getElementPropertyStyle } from '../element/util'
import * as gridPropertyStyleMap from '../element/style'

export interface GridColumn {
  /**
   * The size to assume on this breakpoint. If undefined, the breakpoint will
   * assume the size of the next lowest supplied size, or equal.
   *
   * @see https://getbootstrap.com/docs/4.0/layout/grid/#auto-layout-columns
   * @default equal
   */
  size: BreakpointVal
}

export interface ColumnProps {
  /**
   * The column's extra small / default size.
   *
   * @see https://getbootstrap.com/docs/4.0/layout/grid/#auto-layout-columns
   * @default equal
   */
  size: BreakpointVal
  /**
   * Either the size of the small breakpoint, or an object describing the
   * breakpoint's size and/or styling properties.
   */
  sm: BreakpointPropVal
  /**
   * Either the size of the medium breakpoint, or an object describing the
   * breakpoint's size and/or styling properties.
   */
  md: BreakpointPropVal
  /**
   * Either the size of the large breakpoint, or an object describing the
   * breakpoint's size and/or styling properties.
   */
  lg: BreakpointPropVal
  /**
   * Either the size of the extra-large breakpoint, or an object describing the
   * breakpoint's size and/or styling properties.
   */
  xl: BreakpointPropVal
}

export type BreakpointSize = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12
export type BreakpointVal = 'equal' | 'auto' | string | BreakpointSize
export type BreakpointPropVal = BreakpointVal | GridElement<GridColumn>

export const Col = configureElement<ColumnProps>(
  ({ ownProps, themeProps }) => {
    if (!ownProps.size) {
      ownProps.size = 'equal'
    }

    return Object.keys(ownProps).reduce(
      (style, breakpointId) =>
        assignBreakpointStyle(
          ownProps,
          themeProps,
          breakpointId as keyof ColumnProps,
          style
        ),
      initialColStyle(themeProps)
    )
  },
  ['size', 'sm', 'md', 'lg', 'xl']
)

export function assignBreakpointStyle(
  props: Partial<ColumnProps>,
  theme: ThemeOptions,
  breakpointId: keyof ColumnProps,
  style: CSSObject
) {
  const breakpoint = props[breakpointId]
  const size = typeof breakpoint === 'object' ? breakpoint.size : breakpoint
  const baseStyle = generateBreakpointStyle(breakpointId, size)

  if (typeof breakpoint === 'object') {
    Object.keys(breakpoint).forEach((key) => {
      Object.assign(
        baseStyle,
        getElementPropertyStyle(gridPropertyStyleMap, key, breakpoint[key])
      )
    })
  }

  return Object.assign(
    style,
    breakpointId === 'size'
      ? baseStyle
      : {
          [`@media (min-width: ${theme[breakpointId]}px)`]: baseStyle
        }
  )
}

export function generateBreakpointStyle(
  breakpointId: keyof ColumnProps,
  size: BreakpointVal
) {
  return !size || size === 'equal'
    ? !size && breakpointId !== 'size'
      ? {}
      : breakpointStyleEqual
    : size === 'auto'
    ? breakpointStyleAuto
    : breakpointStyle(size)
}

export function initialColStyle({ spacing }: ThemeOptions): CSSObject {
  const padding = `${spacing}px`
  return {
    width: '100%',
    position: 'relative',
    paddingRight: padding,
    paddingLeft: padding
  }
}

export function breakpointStyle(breakpointVal: string | number): CSSObject {
  const size =
    typeof breakpointVal === 'string'
      ? breakpointVal
      : `${((100 / 12) * breakpointVal).toFixed(6)}%`
  return {
    flex: `0 0 ${size}`,
    maxWidth: `${size}`
  }
}

export const breakpointStyleAuto: CSSObject = {
  width: 'auto',
  maxWidth: '100%',
  flex: '0 0 auto'
}

export const breakpointStyleEqual: CSSObject = {
  maxWidth: '100%',
  flexBasis: 0,
  flexGrow: 1
}
