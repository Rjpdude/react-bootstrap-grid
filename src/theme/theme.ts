import { createElement, createContext, PropsWithChildren } from 'react'
import { ThemeOptions, getDefaultTheme } from './options'
import { GridThemeStyle } from './style'

export const gridThemeContext = createContext(getDefaultTheme())

export function GridTheme(props: PropsWithChildren<Partial<ThemeOptions>>) {
  const style = createElement(GridThemeStyle)
  const context = {
    value: getMergedTheme(props)
  }
  return createElement(
    gridThemeContext.Provider,
    context,
    props.themeStyle !== false && style,
    props.children
  )
}

export function getMergedTheme(themeProps: Partial<ThemeOptions>) {
  const theme = getDefaultTheme()

  Object.keys(theme).forEach((key) => {
    if (themeProps[key]) {
      Object.assign(theme, {
        [key]: themeProps[key]
      })
    }
  })

  return theme
}
