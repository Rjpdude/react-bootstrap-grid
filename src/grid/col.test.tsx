import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import 'jest-styled-components'
import { GridTheme } from '../theme/theme'
import { Col } from './col'

test('Enzyme snapshot', () => {
  const output = mount(
    <Col size={12} id="custom-id" className="custom-class-name">
      <p>child</p>
    </Col>
  )
  expect(toJson(output)).toMatchSnapshot()
})

test('Default col (equal)', () => {
  const output = renderer
    .create(
      <Col>
        <p>child</p>
      </Col>
    )
    .toJSON()

  expect(output).toMatchSnapshot()
  expect(output).toHaveStyleRule('max-width', '100%')
  expect(output).toHaveStyleRule('flex-basis', '0')
  expect(output).toHaveStyleRule('flex-grow', '1')
})

test('Default col (auto)', () => {
  const output = renderer
    .create(
      <Col size="auto">
        <p>child</p>
      </Col>
    )
    .toJSON()

  expect(output).toMatchSnapshot()
  expect(output).toHaveStyleRule('width', 'auto')
  expect(output).toHaveStyleRule('max-width', '100%')
  expect(output).toHaveStyleRule('flex', '0 0 auto')
})

test('Accepts string based column size values', () => {
  const output = renderer
    .create(
      <Col size="20%">
        <p>child</p>
      </Col>
    )
    .toJSON()

  expect(output).toMatchSnapshot()
  expect(output).toHaveStyleRule('flex', '0 0 20%')
  expect(output).toHaveStyleRule('max-width', '20%')
})

test('Breakpoint media', () => {
  const output = renderer
    .create(
      <Col size={12} sm={10} md={8} lg={6} xl={4}>
        <p>child</p>
      </Col>
    )
    .toJSON()

  expect(output).toMatchSnapshot()
  expect(output).toHaveStyleRule('flex', '0 0 100.000000%')
  expect(output).toHaveStyleRule('max-width', '100.000000%')

  // small
  expect(output).toHaveStyleRule('flex', '0 0 83.333333%', {
    media: '(min-width:576px)'
  })
  expect(output).toHaveStyleRule('max-width', '83.333333%', {
    media: '(min-width:576px)'
  })
  // medium
  expect(output).toHaveStyleRule('flex', '0 0 66.666667%', {
    media: '(min-width:768px)'
  })
  expect(output).toHaveStyleRule('max-width', '66.666667%', {
    media: '(min-width:768px)'
  })
  // large
  expect(output).toHaveStyleRule('flex', '0 0 50.000000%', {
    media: '(min-width:992px)'
  })
  expect(output).toHaveStyleRule('max-width', '50.000000%', {
    media: '(min-width:992px)'
  })
  // extra-large
  expect(output).toHaveStyleRule('flex', '0 0 33.333333%', {
    media: '(min-width:1200px)'
  })
  expect(output).toHaveStyleRule('max-width', '33.333333%', {
    media: '(min-width:1200px)'
  })
})

test('Breakpoint media sizing from nested element obj', () => {
  const output = renderer
    .create(
      <Col sm={{ size: 10 }} md={{ size: 8 }} lg={{ size: 6 }} xl={{ size: 4 }}>
        <p>child</p>
      </Col>
    )
    .toJSON()

  expect(output).toMatchSnapshot()

  // small
  expect(output).toHaveStyleRule('flex', '0 0 83.333333%', {
    media: '(min-width:576px)'
  })
  expect(output).toHaveStyleRule('max-width', '83.333333%', {
    media: '(min-width:576px)'
  })
  // medium
  expect(output).toHaveStyleRule('flex', '0 0 66.666667%', {
    media: '(min-width:768px)'
  })
  expect(output).toHaveStyleRule('max-width', '66.666667%', {
    media: '(min-width:768px)'
  })
  // large
  expect(output).toHaveStyleRule('flex', '0 0 50.000000%', {
    media: '(min-width:992px)'
  })
  expect(output).toHaveStyleRule('max-width', '50.000000%', {
    media: '(min-width:992px)'
  })
  // extra-large
  expect(output).toHaveStyleRule('flex', '0 0 33.333333%', {
    media: '(min-width:1200px)'
  })
  expect(output).toHaveStyleRule('max-width', '33.333333%', {
    media: '(min-width:1200px)'
  })
})

test('Applies padding and breakpoints from context', () => {
  const output = renderer
    .create(
      <GridTheme
        spacing={10}
        sm={1}
        smWidth={2}
        md={3}
        mdWidth={4}
        lg={5}
        lgWidth={6}
        xl={7}
        xlWidth={8}
      >
        <Col size={12} sm={10} md={8} lg={6} xl={4}>
          <p>child</p>
        </Col>
      </GridTheme>
    )
    .toJSON()

  expect(output).toMatchSnapshot()
  expect(output).toHaveStyleRule('padding-left', '10px')
  expect(output).toHaveStyleRule('padding-right', '10px')

  // small
  expect(output).toHaveStyleRule('flex', expect.any(String), {
    media: '(min-width:1px)'
  })
  // medium
  expect(output).toHaveStyleRule('flex', expect.any(String), {
    media: '(min-width:3px)'
  })
  // large
  expect(output).toHaveStyleRule('flex', expect.any(String), {
    media: '(min-width:5px)'
  })
  // extra-large
  expect(output).toHaveStyleRule('flex', expect.any(String), {
    media: '(min-width:7px)'
  })
})

test('Applies element style properties on root column', () => {
  const output = renderer
    .create(
      <Col
        p={16}
        justifyContent="between"
        alignSelf="center"
        alignContent="stretch"
      >
        <p>child</p>
      </Col>
    )
    .toJSON()

  expect(output).toMatchSnapshot()
  expect(output).toHaveStyleRule('padding', '16px')
  expect(output).toHaveStyleRule('justify-content', 'space-between')
  expect(output).toHaveStyleRule('align-self', 'center')
  expect(output).toHaveStyleRule('align-content', 'stretch')
})

test('Applies element style properties on breakpoints', () => {
  const output = renderer
    .create(
      <Col
        md={{
          p: 16,
          justifyContent: 'between',
          alignSelf: 'center',
          alignContent: 'stretch'
        }}
      >
        <p>child</p>
      </Col>
    )
    .toJSON()

  expect(output).toMatchSnapshot()

  expect(output).toHaveStyleRule('padding', '16px', {
    media: '(min-width:768px)'
  })
  expect(output).toHaveStyleRule('justify-content', 'space-between', {
    media: '(min-width:768px)'
  })
  expect(output).toHaveStyleRule('align-self', 'center', {
    media: '(min-width:768px)'
  })
  expect(output).toHaveStyleRule('align-content', 'stretch', {
    media: '(min-width:768px)'
  })
})
