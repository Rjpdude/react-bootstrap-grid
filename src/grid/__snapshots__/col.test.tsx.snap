// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accepts string based column size values 1`] = `
.c0 {
  width: 100%;
  position: relative;
  padding-right: 15px;
  padding-left: 15px;
  -webkit-flex: 0 0 20%;
  -ms-flex: 0 0 20%;
  flex: 0 0 20%;
  max-width: 20%;
}

<div
  className="c0 twbs-grid-col"
>
  <p>
    child
  </p>
</div>
`;

exports[`Applies element style properties on breakpoints 1`] = `
.c0 {
  width: 100%;
  position: relative;
  padding-right: 15px;
  padding-left: 15px;
  max-width: 100%;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

@media (min-width:768px) {
  .c0 {
    padding: 16px;
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    -ms-flex-pack: justify;
    justify-content: space-between;
    -webkit-align-self: center;
    -ms-flex-item-align: center;
    align-self: center;
    -webkit-align-content: stretch;
    -ms-flex-line-pack: stretch;
    align-content: stretch;
  }
}

<div
  className="c0 twbs-grid-col"
>
  <p>
    child
  </p>
</div>
`;

exports[`Applies element style properties on root column 1`] = `
.c0 {
  width: 100%;
  position: relative;
  padding-right: 15px;
  padding-left: 15px;
  max-width: 100%;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  padding: 16px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
}

<div
  className="c0 twbs-grid-col"
>
  <p>
    child
  </p>
</div>
`;

exports[`Applies padding and breakpoints from context 1`] = `
.c0 {
  width: 100%;
  position: relative;
  padding-right: 10px;
  padding-left: 10px;
  -webkit-flex: 0 0 100.000000%;
  -ms-flex: 0 0 100.000000%;
  flex: 0 0 100.000000%;
  max-width: 100.000000%;
}

@media (min-width:1px) {
  .c0 {
    -webkit-flex: 0 0 83.333333%;
    -ms-flex: 0 0 83.333333%;
    flex: 0 0 83.333333%;
    max-width: 83.333333%;
  }
}

@media (min-width:3px) {
  .c0 {
    -webkit-flex: 0 0 66.666667%;
    -ms-flex: 0 0 66.666667%;
    flex: 0 0 66.666667%;
    max-width: 66.666667%;
  }
}

@media (min-width:5px) {
  .c0 {
    -webkit-flex: 0 0 50.000000%;
    -ms-flex: 0 0 50.000000%;
    flex: 0 0 50.000000%;
    max-width: 50.000000%;
  }
}

@media (min-width:7px) {
  .c0 {
    -webkit-flex: 0 0 33.333333%;
    -ms-flex: 0 0 33.333333%;
    flex: 0 0 33.333333%;
    max-width: 33.333333%;
  }
}

<div
  className="c0 twbs-grid-col"
>
  <p>
    child
  </p>
</div>
`;

exports[`Breakpoint media 1`] = `
.c0 {
  width: 100%;
  position: relative;
  padding-right: 15px;
  padding-left: 15px;
  -webkit-flex: 0 0 100.000000%;
  -ms-flex: 0 0 100.000000%;
  flex: 0 0 100.000000%;
  max-width: 100.000000%;
}

@media (min-width:576px) {
  .c0 {
    -webkit-flex: 0 0 83.333333%;
    -ms-flex: 0 0 83.333333%;
    flex: 0 0 83.333333%;
    max-width: 83.333333%;
  }
}

@media (min-width:768px) {
  .c0 {
    -webkit-flex: 0 0 66.666667%;
    -ms-flex: 0 0 66.666667%;
    flex: 0 0 66.666667%;
    max-width: 66.666667%;
  }
}

@media (min-width:992px) {
  .c0 {
    -webkit-flex: 0 0 50.000000%;
    -ms-flex: 0 0 50.000000%;
    flex: 0 0 50.000000%;
    max-width: 50.000000%;
  }
}

@media (min-width:1200px) {
  .c0 {
    -webkit-flex: 0 0 33.333333%;
    -ms-flex: 0 0 33.333333%;
    flex: 0 0 33.333333%;
    max-width: 33.333333%;
  }
}

<div
  className="c0 twbs-grid-col"
>
  <p>
    child
  </p>
</div>
`;

exports[`Breakpoint media sizing from nested element obj 1`] = `
.c0 {
  width: 100%;
  position: relative;
  padding-right: 15px;
  padding-left: 15px;
  max-width: 100%;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

@media (min-width:576px) {
  .c0 {
    -webkit-flex: 0 0 83.333333%;
    -ms-flex: 0 0 83.333333%;
    flex: 0 0 83.333333%;
    max-width: 83.333333%;
  }
}

@media (min-width:768px) {
  .c0 {
    -webkit-flex: 0 0 66.666667%;
    -ms-flex: 0 0 66.666667%;
    flex: 0 0 66.666667%;
    max-width: 66.666667%;
  }
}

@media (min-width:992px) {
  .c0 {
    -webkit-flex: 0 0 50.000000%;
    -ms-flex: 0 0 50.000000%;
    flex: 0 0 50.000000%;
    max-width: 50.000000%;
  }
}

@media (min-width:1200px) {
  .c0 {
    -webkit-flex: 0 0 33.333333%;
    -ms-flex: 0 0 33.333333%;
    flex: 0 0 33.333333%;
    max-width: 33.333333%;
  }
}

<div
  className="c0 twbs-grid-col"
>
  <p>
    child
  </p>
</div>
`;

exports[`Default col (auto) 1`] = `
.c0 {
  width: auto;
  position: relative;
  padding-right: 15px;
  padding-left: 15px;
  max-width: 100%;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
}

<div
  className="c0 twbs-grid-col"
>
  <p>
    child
  </p>
</div>
`;

exports[`Default col (equal) 1`] = `
.c0 {
  width: 100%;
  position: relative;
  padding-right: 15px;
  padding-left: 15px;
  max-width: 100%;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

<div
  className="c0 twbs-grid-col"
>
  <p>
    child
  </p>
</div>
`;

exports[`Enzyme snapshot 1`] = `
.c0 {
  width: 100%;
  position: relative;
  padding-right: 15px;
  padding-left: 15px;
  -webkit-flex: 0 0 100.000000%;
  -ms-flex: 0 0 100.000000%;
  flex: 0 0 100.000000%;
  max-width: 100.000000%;
}

<Styled(Col)
  className="custom-class-name"
  id="custom-id"
  size={12}
>
  <Col
    className="custom-class-name twbs-grid-col"
    id="custom-id"
    size={12}
  >
    <styled.div
      className="custom-class-name twbs-grid-col"
      elementProps={Object {}}
      id="custom-id"
      ownProps={
        Object {
          "size": 12,
        }
      }
      themeProps={
        Object {
          "lg": 992,
          "lgWidth": 960,
          "md": 768,
          "mdWidth": 720,
          "sm": 576,
          "smWidth": 540,
          "spacing": 15,
          "themeStyle": true,
          "xl": 1200,
          "xlWidth": 1140,
        }
      }
    >
      <div
        className="c0 custom-class-name twbs-grid-col"
        id="custom-id"
      >
        <p>
          child
        </p>
      </div>
    </styled.div>
  </Col>
</Styled(Col)>
`;
